src/transaction.js

```diff
@@ -327,37 +408,92 @@ Transaction.prototype.byteLength = function () {
   return this.__byteLength(true)
 }
 
-Transaction.prototype.joinsplitByteLength = function () {
-  if (!this.overwintered && this.version < 2) {
+Transaction.prototype.getShieldedSpendByteLength = function () {
+  var byteLength = 0
+  byteLength += varuint.encodingLength(this.vShieldedSpend.length)  // nShieldedSpend
+  byteLength += (384 * this.vShieldedSpend.length)  // vShieldedSpend
+  return byteLength
+}
+
+Transaction.prototype.getShieldedOutputByteLength = function () {
+  var byteLength = 0
+  byteLength += varuint.encodingLength(this.vShieldedOutput.length)  // nShieldedOutput
+  byteLength += (948 * this.vShieldedOutput.length)  // vShieldedOutput
+  return byteLength
+}
+
+Transaction.prototype.getJoinSplitByteLength = function () {
+  // TODO: Throw if we are using the wrong network
+  if (!coins.isZcash(this.network)) {
     return 0
   }
 
-  if (!coins.isZcash(this.network)) {
+  if (this.version < Transaction.ZCASH_JOINSPLITS_SUPPORT_VERSION) {
     return 0
   }
 
   var pubkeySigLength = (this.joinsplits.length > 0) ? (32 + 64) : 0
+
   return (
     bufferutils.varIntSize(this.joinsplits.length) +
     this.joinsplits.reduce(function (sum, joinsplit) {
-      return (
-        sum +
-        8 + 8 + 32 +
-        joinsplit.nullifiers.length * 32 +
-        joinsplit.commitments.length * 32 +
-        32 + 32 +
-        joinsplit.macs.length * 32 +
-        65 + 33 * 7 +
-        joinsplit.ciphertexts.length * Transaction.ZCASH_NOTECIPHERTEXT_SIZE
-      )
+      var byteLength = 0
```

Comment:
Can we simplify this similar to what is done in `getShieldedByteLength` or `getShieldedOutputByteLength`? According to [p78](https://github.com/zcash/zips/blob/master/protocol/protocol.pdf) version 2 and 3 JoinSplits are 1802 bytes and version 4 JoinSplits are 1698 bytes.

Reply:
+1

Reply:
tl;dr: code is the enemy, we should write less code every possible chance we get.

I think we use the numbers Paddy suggested because:

1) more code is almost never good, especially when it replicates what already exists. There is already documentation for the protocol (pg 78 of the PDF added at commit https://github.com/zcash/zips/commit/88e255b63ff96947fdcd5fb0d7d3bc16f0e5b809). We can use the static number and include a link to this commit for interested readers

2). same argument as above

3) same argument as above; if these components have a fixed size then shield txs will probably as well.